lib/logstash/shutdown_controller.rb

```diff
@@ -0,0 +1,98 @@
+# encoding: utf-8
+require 'pp'
+
+module LogStash
+  class ShutdownController
+
+    CHECK_EVERY = 1 # second
+    REPORT_EVERY = 5 # checks
+    ABORT_AFTER = 3 # stalled reports
+    REPORTS = []
+
+    def self.force_shutdown=(boolean)
+      @force_shutdown = boolean
+    end
+
+    def self.force_shutdown?
+      @force_shutdown
+    end
+
+    def logger=(logger)
+      @logger = logger
+    end
+
+    def logger
+      @logger ||= Cabin::Channel.get(LogStash)
+    end
+
+    def initialize(pipeline)
+      @pipeline = pipeline
+    end
+
+    def start(cycle_period=CHECK_EVERY, report_every=REPORT_EVERY, abort_threshold=ABORT_AFTER)
+      @thread ||= Thread.new do
+        sleep(cycle_period)
+        cycle_number = 0
+        stalled_count = 0
+        Stud.interval(cycle_period) do
+          REPORTS << generate_report(@pipeline)
+          REPORTS.delete_at(0) if REPORTS.size > REPORT_EVERY # expire old report
+          if cycle_number == (REPORT_EVERY - 1) # it's report time!
+            logger.warn(REPORTS.last)
+
+            if stalled?
+              logger.error("The shutdown process appears to be stalled due to busy or blocked plugins. Check the logs for more information.") if stalled_count == 0
+              stalled_count += 1
+
+              if self.class.force_shutdown? && abort_threshold == stalled_count
+                logger.fatal("Forcefully quitting logstash..")
+                @pipeline.force_exit()
```

Comment:
I know the implementation details is `exit(-1)`, but the concept is forcing an unsafe shutdown?